rust/arrow/src/json/writer.rs

```diff
@@ -0,0 +1,635 @@
+
+use crate::array::*;
+use crate::datatypes::*;
+use crate::error::Result;
+use crate::record_batch::RecordBatch;
+
+fn primitive_array_to_json<T: ArrowPrimitiveType>(array: &ArrayRef) -> Vec<Value> {
+    as_primitive_array::<T>(array)
+        .iter()
+        .map(|maybe_value| match maybe_value {
+            Some(v) => v.into_json_value().unwrap_or(Value::Null),
+            None => Value::Null,
+        })
+        .collect()
+}
+
+fn struct_array_to_jsonmap_array(
+    array: &StructArray,
+    row_count: usize,
+) -> Vec<JsonMap<String, Value>> {
+    let inner_col_names = array.column_names();
+
+    let mut inner_objs = iter::repeat(JsonMap::new())
+        .take(row_count)
+        .collect::<Vec<JsonMap<String, Value>>>();
+
+    array
+        .columns()
+        .iter()
+        .enumerate()
+        .for_each(|(j, struct_col)| {
+            set_column_for_json_rows(
+                &mut inner_objs,
+                row_count,
+                struct_col,
+                inner_col_names[j],
+            );
+        });
+
+    inner_objs
+}
+
+pub fn array_to_json_array(array: &ArrayRef) -> Vec<Value> {
+    match array.data_type() {
+        DataType::Null => iter::repeat(Value::Null).take(array.len()).collect(),
+        DataType::Boolean => as_boolean_array(array)
+            .iter()
+            .map(|maybe_value| match maybe_value {
+                Some(v) => v.into(),
+                None => Value::Null,
+            })
+            .collect(),
+
+        DataType::Utf8 => as_string_array(array)
+            .iter()
+            .map(|maybe_value| match maybe_value {
+                Some(v) => v.into(),
+                None => Value::Null,
+            })
+            .collect(),
+        DataType::Int8 => primitive_array_to_json::<Int8Type>(array),
+        DataType::Int16 => primitive_array_to_json::<Int16Type>(array),
+        DataType::Int32 => primitive_array_to_json::<Int32Type>(array),
+        DataType::Int64 => primitive_array_to_json::<Int64Type>(array),
+        DataType::UInt8 => primitive_array_to_json::<UInt8Type>(array),
+        DataType::UInt16 => primitive_array_to_json::<UInt16Type>(array),
+        DataType::UInt32 => primitive_array_to_json::<UInt32Type>(array),
+        DataType::UInt64 => primitive_array_to_json::<UInt64Type>(array),
+        DataType::Float32 => primitive_array_to_json::<Float32Type>(array),
+        DataType::Float64 => primitive_array_to_json::<Float64Type>(array),
+        DataType::List(_) => as_list_array(array)
+            .iter()
+            .map(|maybe_value| match maybe_value {
+                Some(v) => Value::Array(array_to_json_array(&v)),
+                None => Value::Null,
+            })
+            .collect(),
+        DataType::Struct(_) => {
+            let jsonmaps =
+                struct_array_to_jsonmap_array(as_struct_array(array), array.len());
+            jsonmaps.into_iter().map(Value::Object).collect()
+        }
+        _ => {
```

Comment:
I think there'd be demand for dictionary support, but we can add that separately

Reply:
Yes I agree -- on both counts :)

Reply:
yeah, i was planning to add dictionary support, but this patch set already grew larger than I expected, so I decided to leave that to a follow up PR for easier review.